client/python/_fastrand.c

```diff
@@ -36,10 +36,12 @@ limitations under the License.
 
 uint64_t randbits(float p1, int num_bits) {
   uint64_t result = 0;
-  int threshold = (int)(p1 * RAND_MAX);
+  uint64_t threshold = p1 * RAND_MAX;
   int i;
   for (i = 0; i < num_bits; ++i) {
-    uint64_t bit = (rand() < threshold);
+    // NOTE: The comparison is <= so that p1 = 1.0 implies that the bit is
```

Comment:
It clearly is not ALL ones. There should be no way for this function to
return all ones. If it can it is a bug.

On Fri, Jan 8, 2016 at 4:25 PM, andychu <notifications@github.com> wrote:

> In client/python/_fastrand.c
> <https://github.com/google/rappor/pull/62#discussion_r49254687>:
>
> >    int i;
> >    for (i = 0; i < num_bits; ++i) {
> > -    uint64_t bit = (rand() < threshold);
> > +    // NOTE: The comparison is <= so that p1 = 1.0 implies that the bit is
>
> Not sure what you're saying here. Did you look at the tests? You get all
> 0's if p = 0.0, and all 1's if p = 1.0. If the comparison were < rather
> than <=, then you would have an edge case when rand() returns RAND_MAX.
> RAND_MAX < RAND_MAX would fail and you would get 0 instead of 1. The idea
> is to have it all be 1's. Andy
> … <#2080025063_>
> On Fri, Jan 8, 2016 at 4:16 PM, Tom Kaitchuck <notifications@github.com>
> wrote: In client/python/_fastrand.c <#62 (comment)
> <https://github.com/google/rappor/pull/62#discussion_r49254217>>: > int
> i; > for (i = 0; i < num_bits; ++i) { > - uint64_t bit = (rand() <
> threshold); > + // NOTE: The comparison is <= so that p1 = 1.0 implies that
> the bit is This comment does not reflect what the change to the code is
> doing, and both are wrong. p=1.0 implies that the value is always zero. I
> think you meant to type p=0 implies the bit is always one. That would be
> true given the change you made. However it is undesirable. We don't want it
> to be the case that if someone sets p=0 they get a bit array of all 1s. It
> should be zeros. Less than is correct. — Reply to this email directly or
> view it on GitHub <
> https://github.com/google/rappor/pull/62/files#r49254217>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/rappor/pull/62/files#r49254687>.
>